pkg/flashdb/mtd/fal_mtd_port.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2022 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup pkg_flashdb
+ * @{
+ *
+ * @file
+ * @brief   MTD <-> FAL adapter
+ *
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @]
+ */
+
+#include "mtd.h"
+
+#include <string.h>
+#include <limits.h>
+#include <fal.h>
+
+#define MIN(a, b) ((a) > (b) ? (b) : (a))
+
+static mtd_dev_t *_mtd;
+
+static int _read(long offset, uint8_t *buf, size_t size)
+{
+    int res = mtd_read_page(_mtd, buf,
+                            offset / _mtd->page_size,
+                            offset % _mtd->page_size,
+                            size);
+    return res ? res : (int)size;
+}
+
+static int _write(long offset, const uint8_t *buf, size_t size)
+{
+    int res = mtd_write_page_raw(_mtd, buf,
+                                 offset / _mtd->page_size,
+                                 offset % _mtd->page_size, size);
+    return res ? res : (int)size;
+}
+
+static int _erase(long offset, size_t size)
+{
+    int res = mtd_erase(_mtd, offset, size);
+    return res ? res : (int)size;
+}
+
+static int _init(void)
+{
+    unsigned sector_size = _mtd->page_size * _mtd->pages_per_sector;
+
+    mtd_flash0.len          = _mtd->sector_count * sector_size;
+    mtd_flash0.blk_size     = sector_size;
+    /* write_gran is in bit, write_size in bytes */
+    mtd_flash0.write_gran   = _mtd->write_size == 1
+                            ? 1
+                            : (_mtd->write_size * 8);
```

Comment:
The documentation says 

```
/* the flash write granularity, unit: bit
- * only support 1(nor flash)/ 8(stm32f2/f4)/ 32(stm32f1) */
```

Should we have an assert to ensure that we are in spec of flashdb?

and isn't ` _mtd->write_size  ^= to write_gran=8`
```suggestion
    mtd_flash0.write_gran   = _mtd->write_size == 1 ?
                            8 :
                            _mtd->write_size == 4 ? 
                            32 : 
                            (assert( _mtd->write_size == 1||_mtd->write_size == 4),0);
```

Reply:
according to the flashdb documentation the stm32f2 requires `write gran = 8` but we can not set this atm s`ince _mtd->write_size == 1`  would be `write gran 8` but is set to 1

Reply:
Yea that's a bit odd, I don't quite understand how write granularity of 1 bit works.
You don't even have 1 bit writes in RAM unless you do bit-banding.

Reply:
i think its not about the write size but the the size of the modification erases flash byte starts at `b11111111` with `write gran 1` next state might be  `b1111 1110` and the next  `b1111 1100` but with write gran 8 the second step is not allowed  therefor the db has to use the next byte.

Reply:
i think its not about the write size but the the size of the modification erases flash byte starts at `b11111111` with `write gran 1` next state might be  `b1111 1110` and the next  `b1111 1100` but with write gran 8 the second step is not allowed  therefor the db has to use the next byte.

<RIOT>/build/pkg/flashdb/src/fdb_utils.c :L95 has a table that might explain that

I am not sure how to deal with that `write gran 1` has wear leveling advantages but we don't know weather the mtd chosen has byte granularity for its modifies or bit (in both cases we would have write_size 1)
This i especially tricky  since one of the examples given for 8 are the stm32f2/f4  and we got a good amount of the nucleos in our boards fall in that range.

8 would be a safe fallback. (the flashdb falls back to higher if two flash are used with different granularity)

i also saw that they themself make use of other granularities that 1/8/32 ->


```suggestion
    mtd_flash0.write_gran   = _mtd->write_size * 8;
```

and maybe some mechanism to overwrite that

Reply:
Looking though the code of FlashDB I can't even find a place where `write_gran` is actually used outside a demo :exploding_head:

Reply:
Seems a bit like the struct member was a red herring the code for the stauts writes makes use of the define. 

[e.g.: here](https://github.com/armink/FlashDB/blob/master/src/fdb_utils.c#L155)

Is that define `FDB_WRITE_GRAN` set by this pkg ?

Reply:
Hm I overlooked that too. How would you suggest to set this? It has to be set at compile time - right now it's just always the default value unless set by the user.

Reply:
Ok, I've set it to 1, `FDB_WRITE_GRAN` is already 1 by default.

Reply:
To double check that this is unused (write only value), I removed front the struct an init() and I got no compiler complains